### ISSUE TITLE
[data grid] row virtualization memory leak

### ISSUE THREAD

**Issue by jorgen-wiebe-calian** [mui/mui-x#12006]
### Steps to reproduce

Link to live example: 
- [Example](https://codesandbox.io/p/sandbox/mui-x-data-grid-memory-leak-example-8jgvjz)
   - I wasn't able to reliably reproduce the bug with the environment provided by the code sandbox, but was able to consistently reproduce the bug if I downloaded the code in the sandbox and ran it locally
Steps:
  - This code example creates 100,000 table rows
  - Every 1 second the rows will be copied to a new array and rows will be updated to the new array 
1. Download the code in the code sandbox
2. `npm install`
3. `npm start`
4. Open the application in goggle chrome
5. Open developer tools and navigate to the memory tab
6. Observe the JavaScript VM memory growing indefinitely
 

### Current behavior

The rows array passed to the DataGridPro component are periodically updated. If the page with the DataGridPro component is idle too long then the browser tab crashes.

Taking a heap snapshot I see:
- ![image](https://github.com/mui/mui-x/assets/136107464/e8894e29-e2f7-428b-a154-7599e541b5db)

https://github.com/mui/mui-x/assets/136107464/7ac33fb7-d91b-400c-a717-90c6b7f7bc27

### Expected behavior

I would expect the used memory to stay relatively consistent and not grow indefinitely.

### Context

We use MUI DataGrid to display status information fetched from a server. Periodically we refetch the data from the server and update the rows in the table. Our client often leaves certain pages open for extended periods of time for monitoring purposes.  

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  @mui/envinfo@2.0.18
Ok to proceed? (y) y

  System:
    OS: Linux 6.5 Ubuntu 22.04.3 LTS 22.04.3 LTS (Jammy Jellyfish)
  Binaries:
    Node: 21.5.0 - /usr/bin/node
    npm: 10.2.4 - /usr/bin/npm
    pnpm: Not Found
  Browsers:
    Chrome: 120.0.6099.129
  npmPackages:
    @emotion/react: 11.11.3 => 11.11.3 
    @emotion/styled: 11.11.0 => 11.11.0 
    @mui/base:  5.0.0-beta.36 
    @mui/core-downloads-tracker:  5.15.9 
    @mui/material: 5.15.9 => 5.15.9 
    @mui/private-theming:  5.15.9 
    @mui/styled-engine:  5.15.9 
    @mui/system:  5.15.9 
    @mui/types:  7.2.13 
    @mui/utils:  5.15.9 
    @mui/x-data-grid:  6.19.3 
    @mui/x-data-grid-pro: 6.19.3 => 6.19.3 
    @mui/x-license-pro:  6.10.2 
    @types/react: 18.2.38 => 18.2.38 
    react: 18.2.0 => 18.2.0 
    react-dom: 18.2.0 => 18.2.0 
    typescript: 4.4.4 => 4.4.4 
```
Browser: Google Chrome
</details>

**Search keywords**: memory leak, virtualization, DataGridPro, DataGrid
**Order ID**: 68118

**Comment by michelengelen** [comment]
@romgrk could you take a look please? Thanks! 🙇🏼

**Comment by romgrk** [comment]
I can reproduce the issue but I'm a bit short on bandwidth at the moment, let's prioritize tomorrow.

**Comment by jorgen-wiebe-calian** [comment]
Is there any update on when this will be fixed, and what version the fix would be apart of?

**Comment by romgrk** [comment]
Sorry we've been busy with the v7 release which came out last week. We'll discuss the task tomorrow.

Which version are you on? If it's possible for you to use v7, we would prefer to fix only on v7, but otherwise we can consider releasing a bufix for v6.

**Comment by jorgen-wiebe-calian** [comment]
We're currently using v6. We aren't going to be able to use v7 to get this fix. We need the bugfix for an upcoming major release, and updating to v7 would require too much effort.

Any updates that come out of your discussion tomorrow would be appreciated. Thanks

**Comment by romgrk** [comment]
We'll look into the bug in the next days. Do you have a timeline for your release, so we know how to prioritize?

We prefer to not bugfix on v6 but this is a pretty serious bug so we can make an exception for that.

**Comment by jorgen-wiebe-calian** [comment]
The release will be sometime in April 2024. We also have pre-release testing that needs to be done, if this could be fixed as soon as possible, that would be great.

**Comment by MBilalShafi** [comment]
I investigated a bit on this, as per my initial observation, this issue seems to be related to facebook/react#14380, the garbage collection for the frequent time-based updates is not great, as observed by other users too.

So I compared the deltas before and after the operation (update rows), and a bigger chunk of retained data in subsequent renders was `memoizedProps` and `memoizedState` from `FiberNode` corresponding to the previous renders.

![image](https://github.com/mui/mui-x/assets/12609561/3567a7cb-0c34-4579-bff8-79c083fc9df4)

Since, the data size is fairly big, and the garbage collection is not being carried out that often due to the frequent updates, the heap size keeps growing. I was able to get to 900MB maximum in ~15 tests which was ultimately garbage collected before the next update. It didn't crash the process on my machine, but it's still a fairly large heap size and may cause issues on machines with fewer resources.

The statement mentioned [here](https://developer.chrome.com/docs/devtools/memory-problems/heap-snapshots) seems to support the point that the main bottleneck is on the garbage collection

> Taking a snapshot always starts with garbage collection.

In all my tests, whenever I took a snapshot, the snapshot size turned out to be significantly smaller than the heap size at that time. For example:

https://github.com/mui/mui-x/assets/12609561/dee691f5-c30f-4e6b-8edb-8fd47a10fef2

Also, on clicking the collect garbage button in the Memory tab, the large heap size also seems to be fixed.

Testing [this](https://github.com/facebook/react/issues/14380#issuecomment-493624355) hypothesis was also helpful. I tried updating the interval time to 10000 from 1000 and ran the test a few times for about 10-30 minutes with no abnormal heap size growth.

So the current conclusion would be:
1. Browser is not able to frequently garbage collect the detached variables when the 
2. The Grid is detaching the variables on the next rows pre-process cycle to be garbage collected just fine (as the garbage collector can pick them when run)

I'll investigate more if there's something we could improve on the Grid's part. @jorgen-wiebe-calian Meanwhile, could you try increasing the interval time on your side and report if you can observe an improvement?

CC @romgrk Let me know if you are familiar with a reliable way to trigger an on-demand garbage collection.

**Comment by romgrk** [comment]
No, don't think it's possible. Maybe it would be worth surfacing the problem to crbug.com. Out of curiosity, is it possible to reproduce the problem without the grid, just by storing the rows in a ref in a simple test component? If the problem is reproducible, then we know for sure we can't do anything on the grid side.

**Comment by jorgen-wiebe-calian** [comment]
I can try decreasing the polling in the example I provided and get back to you.

A few things to note about your results. 

1. In the video I attached in the initial issue you can see me manually trigger the GC and the memory drops quite a bit, but a significantly larger amount is still being used then when the page initially loaded.
2. Where you are looking in your heap snapshot is different than where I was looking. My side shows the memory being retained by "closure"

I've seen similar closure issues before with something like
```js
const foo = () => {
  // create some large variable inside `foo`
  const largeVariable = {...};
  bar = () => {
    // ... do some work then recursively call `foo`
    foo();
  };
  setTimeout(bar, 1000);
};
foo();
```
The `setTimeout` isn't that important, it's there to prevent recursion limits on the infinite loop. The main problem is with where `bar` is defined. `bar` will capture `largeVariable` under closure, but when `bar` calls `foo`, `largeVariable` gets created again and now there are two instances of `largeVariable`, one in the current scope, and one under closure. This recursive loop repeats where more and more `largeVariable` remain under closure. This causes a memory leak. The solution to this is to define `bar` outside of `foo`.

**Comment by jorgen-wiebe-calian** [comment]
@MBilalShafi I modified the example I provided to update the rows every 20 seconds instead of every 1 second and am still seeing the memory leak.

The results of my testing have changed from the initial bug report. I am no longer seeing the rows being retained by "closure" and am seeing it all retained in "object" like your testing showed. This is also true with the 1 second row update rate. I haven't updated any npm packages, nor updated the code from when I made the initial bug report. The only thing that was updated was my chrome version.
